test/e2e/apps/deployment.go

```diff
@@ -134,6 +141,238 @@ var _ = SIGDescribe("Deployment", func() {
 	// See https://github.com/kubernetes/kubernetes/issues/29229
+
+	ginkgo.It("should run the lifecycle of a Deployment", func() {
+		deploymentResource := schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments"}
+		testNamespaceName := f.Namespace.Name
+		testDeploymentName := "test-deployment"
+		testDeploymentInitialImage := "nginx"
+		testDeploymentPatchImage := "alpine"
+		testDeploymentUpdateImage := "httpd"
+		testDeploymentDefaultReplicas := int32(3)
+		testDeploymentMinimumReplicas := int32(1)
+		testDeploymentNoReplicas := int32(0)
+		testDeploymentLabels := map[string]string{"test-deployment-static": "true"}
+		testDeploymentLabelsFlat := "test-deployment-static=true"
+		testDeploymentLabelSelectors := metav1.LabelSelector{
+			MatchLabels: testDeploymentLabels,
+		}
+
+		ginkgo.By("creating a Deployment")
+		testDeployment := appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   testDeploymentName,
+				Labels: map[string]string{"test-deployment-static": "true"},
+			},
+			Spec: appsv1.DeploymentSpec{
+				Replicas: &testDeploymentDefaultReplicas,
+				Selector: &testDeploymentLabelSelectors,
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: testDeploymentLabelSelectors.MatchLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{{
+							Name:  testDeploymentName,
+							Image: testDeploymentInitialImage,
+						}},
+					},
+				},
+			},
+		}
+		_, err := f.ClientSet.AppsV1().Deployments(testNamespaceName).Create(context.TODO(), &testDeployment, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "failed to create Deployment %v in namespace %v", testDeploymentName, testNamespaceName)
+
+		ginkgo.By("watching for the Deployment to be added")
+		dplmtWatchTimeoutSeconds := int64(180)
+		dplmtWatch, err := f.ClientSet.AppsV1().Deployments(testNamespaceName).Watch(context.TODO(), metav1.ListOptions{LabelSelector: testDeploymentLabelsFlat, TimeoutSeconds: &dplmtWatchTimeoutSeconds})
+		framework.ExpectNoError(err, "Failed to setup watch on newly created Deployment")
+
+		foundEvent := false
+		dplmtWatchChan := dplmtWatch.ResultChan()
+		for event := range dplmtWatchChan {
+			if event.Type == watch.Added {
+				foundEvent = true
+				break
+			}
+		}
+		framework.ExpectEqual(foundEvent, true, "failed to find watch event %v", watch.Added)
+
+		ginkgo.By("waiting for all Replicas to be Ready")
+		foundEvent = false
+		for event := range dplmtWatchChan {
+			deployment, ok := event.Object.(*appsv1.Deployment)
+			framework.ExpectEqual(ok, true, "unable to convert event.Object type")
+			if deployment.Status.AvailableReplicas == testDeploymentDefaultReplicas &&
+				deployment.Status.ReadyReplicas == testDeploymentDefaultReplicas {
+				foundEvent = true
+				break
+			}
+		}
+		framework.ExpectEqual(foundEvent, true, "failed to see scale of replicas")
```

Comment:
Have you considered using watchtools.Until?

Reply:
Thank you for the suggestion, it's seemed to work really well in other current PRs. I'll adapt it in this test too

Reply:
Updated in fc8e24c6869